File: server/socket/socket.go
package socket

import (
	"encoding/json"

	socketio "github.com/googollee/go-socket.io"
)

type ChatMsg struct {
	Message		string	`json:"message"`
	Author		string	`json:"author"`
}

func SocketSetup(server *socketio.Server) {
	server.On("connection", func(so socketio.Socket) {
		so.On("join-room", func(roomID string) {
			so.Join(roomID)
	
			so.On("greeting-message", func(msg ChatMsg) {
				encodedMsg, _ := json.Marshal(msg)
				server.BroadcastTo(roomID, "greeting-message", string(encodedMsg))
			})
	
			so.On("chat-message", func(msg ChatMsg) {
				encodedMsg, _ := json.Marshal(msg)
				server.BroadcastTo(roomID, "chat-message", string(encodedMsg))
			})
		})
	})
}

File: nodejs-v/server/socket.ts
import { Server, Socket } from 'socket.io';

interface Data {
    message: string,
    author: string
};

export default (io: Server) => {
    io.on('connection', (socket: Socket) => {
        socket.on('join-room', (room: string) => {
            socket.join(room);
            socket.on('greeting-message', (data: Data) => {
                io.to(room).emit('greeting-message', {
                    message: data.message,
                    author: data.author
                });
            });
            socket.on('chat-message', (data: Data) => {
                io.to(room).emit('chat-message', {
                    message: data.message,
                    author: data.author
                });
            });
        });
    });
};
